codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -263,60 +283,71 @@ private void cleanup() {
-            if (clientPrefaceString == null) {
-                return true;
-            }
-
             int prefaceRemaining = clientPrefaceString.readableBytes();
-            int bytesRead = Math.min(in.readableBytes(), prefaceRemaining);
+            int bytesRead = min(in.readableBytes(), prefaceRemaining);
 
             // If the input so far doesn't match the preface, break the connection.
             if (bytesRead == 0 || !ByteBufUtil.equals(in, in.readerIndex(),
                     clientPrefaceString, clientPrefaceString.readerIndex(), bytesRead)) {
-                throw connectionError(PROTOCOL_ERROR, "HTTP/2 client preface string missing or corrupt.");
+                String receivedBytes = hexDump(in.slice(in.readerIndex(),
+                        min(in.readableBytes(), clientPrefaceString.readableBytes())));
+                throw connectionError(PROTOCOL_ERROR, "HTTP/2 client preface string missing or corrupt. " +
+                        "Hex dump for received bytes: %s", receivedBytes);
             }
             in.skipBytes(bytesRead);
             clientPrefaceString.skipBytes(bytesRead);
 
             if (!clientPrefaceString.isReadable()) {
                 // Entire preface has been read.
-                clientPrefaceString.release();
-                clientPrefaceString = null;
+                cleanup();
                 return true;
             }
             return false;
         }
+    }
 
-        /**
-         * Sends the HTTP/2 connection preface upon establishment of the connection, if not already sent.
-         */
-        private void sendPreface(ChannelHandlerContext ctx) {
-            if (prefaceSent || !ctx.channel().isActive()) {
-                return;
-            }
-
-            prefaceSent = true;
-
-            if (!connection().isServer()) {
-                // Clients must send the preface string as the first bytes on the connection.
-                ctx.write(connectionPrefaceBuf()).addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
-            }
+    private final class FrameDecoder extends BaseDecoder {
+        private boolean firstFrame = true;
 
-            // Both client and server must send their initial settings.
-            encoder.writeSettings(ctx, initialSettings(), ctx.newPromise()).addListener(
-                    ChannelFutureListener.CLOSE_ON_FAILURE);
+        FrameDecoder(ChannelHandlerContext ctx, boolean prefaceSent) {
+            super(ctx, prefaceSent);
         }
-    }
 
-    private final class FrameDecoder extends BaseDecoder {
         @Override
-        public void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
             try {
+                // If it's the first frame we've decoded, make sure it's a SETTINGS frame.
+                if (firstFrame && verifyFrameIsSettings(in)) {
```

Comment:
Can we move this logic into `PrefaceDecoder`?  One of the reasons why `FrameDecoder` was introduced was to avoid having a conditional statement evaluate on every `decode` which will only be relevant for the first (or first couple) decodes.